File: common/events.py
"""Événements typés — canal WebSocket et journal d'audit (section 11).

Nomenclature contractuelle consommée par le frontend (websocket_live), les
plugins EDA (kicad_live_host, altium_bridge) et le credit_dashboard :
    net_routed · component_moved · drc_update · step_progress ·
    optimizer_iteration · export_ready · credit_debit · session_resumed

Chaque événement porte : project_id, version du design, type, charge utile
JSON, horodatage monotone et identifiant de l'émetteur — ce qui rend la
reprise par delta possible depuis le state_manager.
"""

from __future__ import annotations

import itertools
import time
import uuid
from dataclasses import dataclass, field
from enum import Enum
from typing import Any

_SEQ = itertools.count()


class EventType(str, Enum):
    NET_ROUTED = "net_routed"
    COMPONENT_MOVED = "component_moved"
    DRC_UPDATE = "drc_update"
    STEP_PROGRESS = "step_progress"
    OPTIMIZER_ITERATION = "optimizer_iteration"
    EXPORT_READY = "export_ready"
    CREDIT_DEBIT = "credit_debit"
    SESSION_RESUMED = "session_resumed"
    # Événements internes du pipeline multi-agents
    PLAN_UPDATED = "plan_updated"
    BOM_VALIDATED = "bom_validated"
    SKIDL_GENERATED = "skidl_generated"
    ERC_ERROR = "erc_error"
    CONSTRAINT_VIOLATED = "constraint_violated"
    ROLLBACK_PERFORMED = "rollback_performed"
    FIRMWARE_REGENERATED = "firmware_regenerated"


class WorkflowStep(int, Enum):
    """Les huit étapes du workflow de données (section 09)."""

    NL_TO_SKIDL = 1          # [Circuitron + Flux]
    MULTI_AGENT_PLANNING = 2 # [Siemens + Circuitron]
    RL_PLACE_ROUTE = 3       # [DeepPCB + Siemens]
    NIGHT_OPTIMIZATION = 4   # [AutoPCB]
    MULTI_PHYSICS_CHECK = 5  # [Cadence AuraStack]
    SURGICAL_EDIT = 6        # [Flux.ai]
    FIRMWARE_GENERATION = 7  # [Flux.ai]
    EXPORT_FEEDBACK = 8      # [DeepPCB + Siemens]

    @property
    def label(self) -> str:
        return _STEP_LABELS[self]


_STEP_LABELS: dict[WorkflowStep, str] = {
    WorkflowStep.NL_TO_SKIDL: "Entrée — langage naturel vers SKiDL",
    WorkflowStep.MULTI_AGENT_PLANNING: "Planification multi-agents",
    WorkflowStep.RL_PLACE_ROUTE: "Placement & routage RL",
    WorkflowStep.NIGHT_OPTIMIZATION: "Optimisation autonome de nuit",
    WorkflowStep.MULTI_PHYSICS_CHECK: "Vérification multi-physique",
    WorkflowStep.SURGICAL_EDIT: "Modifications chirurgicales",
    WorkflowStep.FIRMWARE_GENERATION: "Génération firmware",
    WorkflowStep.EXPORT_FEEDBACK: "Export & rétroaction usine",
}


@dataclass
class Event:
    """Événement typé transmis sur le canal WebSocket / journal d'audit."""

    type: EventType
    project_id: str
    payload: dict[str, Any] = field(default_factory=dict)
    design_version: int = 1
    emitter: str = "system"
    ts: float = field(default_factory=time.time)
    seq: int = field(default_factory=lambda: next(_SEQ))
    event_id: str = field(default_factory=lambda: uuid.uuid4().hex)

    def to_json(self) -> dict[str, Any]:
        return {
            "event_id": self.event_id,
            "seq": self.seq,
            "ts": self.ts,
            "type": self.type.value,
            "project_id": self.project_id,
            "design_version": self.design_version,
            "emitter": self.emitter,
            "payload": self.payload,
        }

    @classmethod
    def from_json(cls, raw: dict[str, Any]) -> "Event":
        return cls(
            type=EventType(raw["type"]),
            project_id=raw["project_id"],
            payload=raw.get("payload", {}),
            design_version=raw.get("design_version", 1),
            emitter=raw.get("emitter", "system"),
            ts=raw.get("ts", time.time()),
            seq=raw.get("seq", 0),
            event_id=raw.get("event_id", uuid.uuid4().hex),
        )


def make_event(
    event_type: EventType | str,
    project_id: str,
    emitter: str = "system",
    design_version: int = 1,
    **payload: Any,
) -> Event:
    """Fabrique un événement : make_event(EventType.NET_ROUTED, pid, "router", net_id=...)"""
    if isinstance(event_type, str):
        event_type = EventType(event_type)
    return Event(
        type=event_type,
        project_id=project_id,
        emitter=emitter,
        design_version=design_version,
        payload=payload,
    )

File: frontend/src/services/websocket_live/events.js
/* Nomenclature contractuelle des événements WebSocket (spécification section 11).
 *
 * Chaque événement circulant sur /ws porte :
 *   { event_id, seq, ts, type, project_id, design_version, emitter, payload }
 * Les types ci-dessous SONT le contrat — ne pas renommer.
 */

export const EVENT_TYPES = {
  NET_ROUTED: 'net_routed',
  COMPONENT_MOVED: 'component_moved',
  DRC_UPDATE: 'drc_update',
  STEP_PROGRESS: 'step_progress',
  OPTIMIZER_ITERATION: 'optimizer_iteration',
  EXPORT_READY: 'export_ready',
  CREDIT_DEBIT: 'credit_debit',
  SESSION_RESUMED: 'session_resumed',
  // Événements internes du pipeline multi-agents (section 06)
  PLAN_UPDATED: 'plan_updated',
  BOM_VALIDATED: 'bom_validated',
  SKIDL_GENERATED: 'skidl_generated',
  ERC_ERROR: 'erc_error',
  CONSTRAINT_VIOLATED: 'constraint_violated',
  ROLLBACK_PERFORMED: 'rollback_performed',
  FIRMWARE_REGENERATED: 'firmware_regenerated',
};

/** Descriptions françaises — affichées dans les toasts et le journal live. */
export const EVENT_DESCRIPTIONS = {
  net_routed: 'Une piste a été routée — tracé animé dans le viewer 3D',
  component_moved: 'Un composant a été déplacé — position tweenée en direct',
  drc_update: 'Mise à jour du score DRC/DFM courant',
  step_progress: 'Progression du pipeline (étape sur 8)',
  optimizer_iteration: 'Itération de l’optimiseur nocturne (AutoPCB)',
  export_ready: 'Artefact d’export disponible (Gerber, ODB++, firmware)',
  credit_debit: 'Débit de crédits pay-as-you-go (brique DeepPCB)',
  session_resumed: 'Session restaurée — re-synchronisation par delta depuis le state_manager',
  plan_updated: 'Plan de conception mis à jour par le planner multi-agents',
  bom_validated: 'BOM validé (stocks, prix, empreintes)',
  skidl_generated: 'Netlist SKiDL générée depuis le langage naturel',
  erc_error: 'Erreur ERC détectée sur le schéma',
  constraint_violated: 'Contrainte de conception violée (bus de contraintes)',
  rollback_performed: 'Retour à une version antérieure du design',
  firmware_regenerated: 'Headers firmware régénérés (Zephyr/Arduino)',
};

/** Libellés courts pour l’UI. */
const _PRETTY = {
  net_routed: 'Routage de piste',
  component_moved: 'Déplacement de composant',
  drc_update: 'Score DRC',
  step_progress: 'Étape du pipeline',
  optimizer_iteration: 'Itération optimiseur',
  export_ready: 'Export prêt',
  credit_debit: 'Débit crédits',
  session_resumed: 'Session restaurée',
  plan_updated: 'Plan mis à jour',
  bom_validated: 'BOM validé',
  skidl_generated: 'SKiDL généré',
  erc_error: 'Erreur ERC',
  constraint_violated: 'Contrainte violée',
  rollback_performed: 'Rollback',
  firmware_regenerated: 'Firmware régénéré',
};

/** Les événements qui modifient le design en direct (< 100 ms visés). */
const _LIVE_TYPES = new Set([
  EVENT_TYPES.NET_ROUTED,
  EVENT_TYPES.COMPONENT_MOVED,
  EVENT_TYPES.DRC_UPDATE,
  EVENT_TYPES.OPTIMIZER_ITERATION,
]);

/** true si l’événement anime le design en temps réel (viewer + éditeur). */
export function isLiveEvent(type) {
  return _LIVE_TYPES.has(type);
}

/** Libellé court français d’un type d’événement. */
export function prettyLabel(type) {
  return _PRETTY[type] || (typeof type === 'string' ? type : 'événement');
}

/** Les 8 étapes du workflow (section 09) — labels français + émetteurs. */
export const WORKFLOW_STEPS = [
  { index: 1, label: 'Langage naturel → SKiDL', emitters: 'Flux.ai · Circuitron' },
  { index: 2, label: 'Planification multi-agents', emitters: 'Siemens Fuse · Circuitron' },
  { index: 3, label: 'Placement & routage RL', emitters: 'DeepPCB · Siemens Fuse' },
  { index: 4, label: 'Optimisation autonome de nuit', emitters: 'AutoPCB' },
  { index: 5, label: 'Vérification multi-physique', emitters: 'Cadence AuraStack' },
  { index: 6, label: 'Modifications chirurgicales', emitters: 'Flux.ai' },
  { index: 7, label: 'Génération firmware', emitters: 'Flux.ai' },
  { index: 8, label: 'Export & rétroaction usine', emitters: 'DeepPCB · Siemens Fuse' },
];

/** Libellé français de l’étape n (1..8), chaîne vide si hors bornes. */
export function stepLabel(stepIndex) {
  const found = WORKFLOW_STEPS.find((s) => s.index === Number(stepIndex));
  return found ? found.label : '';
}
